### PR TITLE
Gem side changes to support Slack integration for Blazer

### PR DESCRIPTION
**Pull request by moelee**
App side PR: https://github.com/instacart/carrot/pull/9492

### What
Add support for posting blazer check notifications to slack.

### Why (optional)
Because notifications in slack would be cool

### Who should review this
@ankane 

### How it looks
image

### TODO
- [x] Test against private channels, individuals. Might only work against public channels or channels that Moses is already in. If public channels only, note that in the form field help text?
- [x] Make slack API call async (suckerpunch?)
- [x] Refactor out code that build and sends slack notification into a service class
- [x] Verify config methodology is sane

### CODE REVIEW

**Comment by moelee** [comment]
Confirmed that Slack webhook allows posting to both private and public channels. w00t!

**Comment by moelee** [comment]
@ankane this is ready for review. See hosting app side changes here also: https://github.com/instacart/carrot/pull/9492

File: lib/blazer/slack_notifier.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      def manage_checks_url
+        @manage_checks_url ||= Blazer::Engine.routes.url_helpers.checks_url
+      end
+
+      def send_state_change(check, result, query, state)
+        query_url = Blazer::Engine.routes.url_helpers.query_url(query)
+        color = STATE_COLOR_MAP[state]
+
+        check.split_slack_channels.each do |slack_channel|
+          json = {
```
Comment (ankane) [inline lib/blazer/slack_notifier.rb]: Let's dry up the payload / HTTP request.

File: app/models/blazer/check.rb [hunk 2]
```diff
@@ -58,7 +63,27 @@ def update_state(result)
 
       # do not notify on creation, except when not passing
       if notify?
-        Blazer::CheckMailer.state_change(self, state, state_was, result.rows.size, message, result.columns, result.rows.first(10).as_json, result.column_types, check_type).deliver_now
+        if emails.present?
+          Blazer::CheckMailer.state_change(self, state, state_was, result.rows.size, message, result.columns, result.rows.first(10).as_json, result.column_types, check_type).deliver_now
+        end
+
+        if slack_channels.present?
+          if Blazer.async
```
Comment (ankane) [inline app/models/blazer/check.rb]: Async just applies to how queries are run, so shouldn't need a `sucker_punch` version.

File: app/views/blazer/checks/index.html.erb [hunk 3]
```diff
@@ -30,6 +31,13 @@
             <% end %>
           </ul>
         </td>
+        <td>
+          <ul class="list-unstyled" style="margin-bottom: 0; word-break: break-all;">
```
Comment (ankane) [inline app/views/blazer/checks/index.html.erb]: Would be good to include this in the emails column (and probably rename it) so things are squished.

**Comment by moelee** [comment]
updated incorporating last round of feedback

File: app/views/blazer/checks/index.html.erb [hunk 4]
```diff
@@ -9,7 +9,7 @@
       <th>Query</th>
       <th style="width: 10%;">State</th>
       <th style="width: 10%;">Run</th>
-      <th style="width: 20%;">Emails</th>
+      <th style="width: 20%;">Emails & Slack Channels</th>
```
Comment (ankane) [inline app/views/blazer/checks/index.html.erb]: Think this should be `&amp;`

**Comment by ankane** [comment]
@moelee Looks good to me. I haven't tested, but feel free to merge if you're feeling good.

**Comment by moelee** [comment]
Double check private channel support against a channel I've never been invited to before and it successfully posted to there.

**Pull request merged**